src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,125 @@ public SubstitutionMatrix(final byte[] matrix) {
         baseByCode['N'][(encodedMatrixBytes[4] >> 2) & 3] = 'G';
         baseByCode['N'][(encodedMatrixBytes[4]) & 3] = 'T';
 
-        for (final byte refBase : SUBSTITUTION_BASES_UPPER) {
+        for (final SubstitutionBase refBase : BASES) {
             for (byte code = 0; code < CODES_PER_BASE; code++) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (refBase <= 0 || Character.isLowerCase((char) refBase)) {
+            throw new IllegalArgumentException(
+                    String.format("CRAM: Attempt to generate a substitution code for invalid or lower case reference base '%c'", (char) refBase));
+        }
+        if (readBase <= 0) {
```

Comment:
I wonder... why the asymmetric treatment for read bases... cannot these also be erroneous (e.g. lowercase or symbols?) why not croak on those with we do it with refBase.